src/main/java/org/truffleruby/core/queue/SizedQueue.java

```diff
@@ -127,6 +127,28 @@ public boolean put(Object item) throws InterruptedException {
         }
     }
 
+    @TruffleBoundary
+    public Object put(Object item, long timeoutMilliseconds) throws InterruptedException {
+        lock.lock();
+        try {
+            if (size == capacity) {
+                final boolean signalled = ConcurrentOperations.awaitAndCheckInterrupt(canAdd, timeoutMilliseconds,
+                        TimeUnit.MILLISECONDS);
+                if (!signalled) {
+                    return false;
+                }
+                if (closed) {
+                    return CLOSED;
+                }
```

Comment:
Should we check this earlier, like in `public boolean put(Object item)`? Or not?
Please make sure there is a spec testing that (add one if there isn't).

Reply:
True. Checked CRuby and it doesn't wait when it's already closed, so an early check will be needed. However one at the end is also required to ensure the queue wasn't closed in the middle of the wait. Adding test too.